doc/decisions/constructor_functions_examples.md

```diff
@@ -0,0 +1,105 @@
+### Constructor Functions: Language examples
+
+This is not a decision.
+This file contains additional information for [Constructor Functions](constructor_functions.md).
+
+## Rust
+
+For example, `libelektra-rust` (targeted at Rust) could provide:
+
+```rust
+trait Key {
+    fn new(ns: Namespace, name: &[u8]) -> Self;
+    fn set_value(&mut self, value: &[u8]) -> Self;
+    fn set_meta(&mut self, name: &[u8], value: &[u8]) -> Self;
+}
+```
+
+In Rust it is possible to access the length of a `&[u8]` (otherwise similar to a `uint8_t *`), therefore the issue of separate `size` arguments doesn't come up.
+
+Similarly, for keysets we could provide:
+
+```rust
+trait KeySet {
+    fn new (keys: &[Key]) -> Self;
+}
+```
+
+## C
+
+For C it is a bit more complicated.
+`libelektra-operations` could provide an API based on variadic arguments:
+
+```c
+// KEY_NAME_END must be NULL, because everything else could in theory be valid char * for a name part
+
+elektraKeyCreate (KEY_NS_USER, "foo", "bar", KEY_NAME_END,
+        KEY_VALUE_STRING, "value",
+        KEY_META, elektraKeyCreate (KEY_NS_META, "type", KEY_NAME_END, KEY_VALUE_STRING, "string", KEY_END),
+        KEY_META, elektraKeyCreate (KEY_NS_META, "other", KEY_NAME_END, KEY_VALUE_BINARY, &mystruct, sizeof(mystruct), KEY_END),
+        KEY_END);
+```
+
+However, this would require multiple `memcpy` to initialize the key name.
+
+Alternatively, macros could be used extensively to hide the size arguments from (most) users:
```

Comment:
I don't think we need this.

Reply:
I'm not a fan of the varargs API there are a few downsides:
- multiple memcpy for the name (could still be avoided with some macros and/or extra arguments)
- it's very easy to pass wrong arguments and get weird errors. For example, if one of the name parts is a variable which is `NULL` for some reason.
- There are basically no compiler checks for a vararg API. With the many macros below, errors may be hard to read, but the compiler will still check the result after expanding the macros, so fewer errors are possible.

Reply:
I think, I have found a compromise. The idea revolves around a public `struct` and a `static inline` function.

```c
// holds a string (with potential zero bytes) of `size` length
// if `ns == ELEKTRA_NS_NONE`, the `name` and `size` are directly useable in struct _Key
// otherwise:
//  - `size` is the length of `name` including the terminating zero byte, but without a namespace
//  - `name` does not contain a namespace
//  - the full name is constructed from `ns` and `name`
struct ElektraKeyName {
    size_t size;
    const char * name;
    ElektraNamespace ns;
};

static inline struct ElektraKeyName elektraKeyName(ElektraNamespace ns, const char * parts[])
{
    struct ElektraKeyName k;
    k.ns = ELEKTRA_NS_NONE;

    k.size = 2;
    for (size_t i = 0; parts[i] != NULL; ++i)
    {
        k.size += strlen(parts[i]) + 1;
    }
          
    char * name = malloc(k.size);
    name[0] = ns;
    name[1] = 0;
    k.size = 2;
    for (size_t i = 0; parts[i] != NULL; ++i)
    {
        size_t s = strlen(parts[i]);
        memcpy(name + k.size, parts[i], s);
        name[k.size + s] = 0;
        k.size += s + 1;
    }
    k.name = name;
    return k;
}
```

This would be part of the header for `libelektra-operations` (or a similar library). This allows us to define a `elektraKeyCreate` function like this:

```c
Key * elektraKeyCreate(struct ElektraKeyName name, void * value, size_t valueSize, KeySet * meta)
```

All of this may seem odd at first, but combining the arguments for namespace, name and size of name into a struct, allows us to create them via macros (which makes the syntax nicer).

```c
#define ELEKTRA_BUILD_NAME(ns, ...) (elektraKeyName(ns, (const char *[]){__VA_ARGS__,NULL}))
#define ELEKTRA_KEY_NAME(_ns, _name, _size) ((struct ElektraKeyName){.size = _size, .name = _name, .ns = _ns})
#define ELEKTRA_LITERAL_NAME(_ns, _name) (ELEKTRA_KEY_NAME(_ns, _name, sizeof(""_name)))
```

All in all the effect is that `elektraKeyCreate` can be called in many different ways:

```c
    struct ElektraKeyName y = ELEKTRA_BUILD_NAME(ELEKTRA_NS_CASCADING, "abc", "def", "xyz", "123", "456", "ghi", "foo", "lll", "kyt");
    Key * k = elektraKeyCreate(y, NULL, 0, NULL);
    
    Key * k1 = elektraKeyCreate(ELEKTRA_BUILD_NAME(ELEKTRA_NS_CASCADING, "abc", "def"), NULL, 0, NULL);
    
    Key * k2 = elektraKeyCreate(ELEKTRA_LITERAL_NAME(ELEKTRA_NS_CASCADING, "foo\0bar"), NULL, 0, NULL);

    const char *x = "foo\0bar";
    Key * k3 = elektraKeyCreate(ELEKTRA_KEY_NAME(ELEKTRA_NS_CASCADING, x, 8), NULL, 0, NULL);

    const char *a = "foo";
    const char *b = "bar";
    Key * k4 = elektraKeyCreate(ELEKTRA_BUILD_NAME(ELEKTRA_NS_CASCADING, "abc", a, b, "def"), NULL, 0, NULL);
```

Crucially when the `static inline elektraKeyName` function is used, we directly use the `elektraMalloc`ed buffer. This in turn makes it possible for the compiler (with `-O3`) to do lots of constant folding and (in theory) fully calculate the key name at compile-time in many instances. After inlining and constant folding, only the `malloc` with a precomputed size and a (vectorized) `memcpy` from static memory to the heap remain.

Sadly, however, it seems GCCs constant folding algorithm isn't powerful enough. But Clang can handle even the more complex cases as this Godbolt link shows: https://godbolt.org/z/Tf1rnxfMW (The important part is that in the assembly, only the last `elektraKeyCreate` is preceeded by a `memcpy` from the inlined function)

Note also, that only the creation of the key name is inlined. The rest is still hidden behind a layer of abstraction, so this does not leak the layout of `struct _Key` to the API consumer. Inlining the allocation of the key name is not a problem, because the structure of the key name is already public API..